From 0f5b6b1d495747e307ce47909f612b58aafb2e50 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Tue, 17 Sep 2013 16:14:17 -0700
Subject: [PATCH] kvm: enable KVM_MIPS_INJECT_PCIE_INTX to inject either INTx
 or MSIX interrupts to guest

Commit bd55482deb66d7c7744911169d9c08c8c4c6cf22 from Broadcom SDK 3.0.2

Enable KVM_MIPS_INJECT_PCIE_INTX to inject either INTx or MSIX
interrupts to guest

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index 10a5459..2859215 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -903,9 +903,19 @@ long kvm_arch_vm_ioctl(struct file *filp, unsigned int ioctl, unsigned long arg)
 	case KVM_MIPS_INJECT_PCIE_INTX:
 	{
 		struct kvm_arch *arch = &kvm->arch;
-		int link = (int)(long)argp;
-
-		kvm_pic_inject_guest_ext(kvm, arch, PIC_XLP9XX_IRT_PCIE_LINK_0_INDEX + link, 0);
+		int type, val = (int)(long)argp;
+
+		type = val >> 16;
+		if (type == 0)
+			/* PCI Link INTx */
+			kvm_pic_inject_guest_ext(kvm, arch,
+				PIC_XLP9XX_IRT_PCIE_LINK_0_INDEX + (val & 0xffff), 0);
+		else if (type == 1)
+			/* PCI MSIX */
+			kvm_pic_inject_guest_ext(kvm, arch,
+				PIC_IRT_PCIE_MSIX_INDEX(val & 0xffff), 0);
+		else
+			r = -EINVAL;
 		break;
 	}
 	default:
-- 
1.9.1

